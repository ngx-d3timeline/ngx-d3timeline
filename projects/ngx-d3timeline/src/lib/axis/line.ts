import { Orientation } from '../orientation';

export interface Line {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
}

export function createLine(
  x1: number,
  y1: number,
  length: number,
  orientation: Orientation
) {
  return orientation === Orientation.Vertical
    ? createVerticalLine(x1, y1, length)
    : createHorizontalLine(x1, y1, length);
}

export function createVerticalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1,
    y2: y1 + length
  };
}

export function createHorizontalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1 + length,
    y2: y1
  };
}
