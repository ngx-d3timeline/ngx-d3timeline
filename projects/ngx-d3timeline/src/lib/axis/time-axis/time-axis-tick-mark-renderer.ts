import { TimeScale } from '../../scale-types';
import { TickMarkRenderer } from '../tick-mark-renderer';

export class TimeAxisTickMarkRenderer implements TickMarkRenderer<TimeScale> {
  getTickValues(scale: TimeScale): Date[] {
    return scale.ticks();
  }

  getTickLabel(scale: TimeScale, tickValue: any): string {
    return scale.tickFormat()(tickValue);
  }

  getTransform(scale: TimeScale, tickValue: any): number {
    return scale(tickValue);
  }

  getTickLineOffset(): number {
    return -5;
  }
}
