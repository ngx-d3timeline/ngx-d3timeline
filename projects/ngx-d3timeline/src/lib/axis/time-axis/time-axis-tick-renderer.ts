import { TimeScale } from '../../scale-types';
import { TickRenderer } from '../tick-renderer';

export class TimeAxisTickRenderer implements TickRenderer<TimeScale> {
  getTickValues(scale: TimeScale): Date[] {
    return scale.ticks();
  }

  getLabel(scale: TimeScale, tick: any): string {
    return scale.tickFormat()(tick);
  }

  getTransform(scale: TimeScale, tick: any): number {
    return scale(tick);
  }

  getTickLineOffset(): number {
    return -5;
  }
}
