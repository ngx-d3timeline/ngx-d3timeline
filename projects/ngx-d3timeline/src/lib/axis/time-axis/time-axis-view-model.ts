import { scaleTime, ScaleTime } from 'd3-scale';
import { max, min } from 'd3-array';
import { TimelineEvent } from '../../timeline-event';
import { TimelineView } from '../../view/timeline-view';
import { Orientation } from '../../orientation';

export class TimeAxisViewModel {
  readonly scaleTime: ScaleTime<number, number>;

  constructor(
    data: TimelineEvent[],
    timelineView: TimelineView,
    event: any,
    private readonly orientation: Orientation
  ) {
    this.scaleTime = this.configureScaleTime(data, timelineView, event);
  }

  get tickInfos(): { label: string; transform: string }[] {
    return this.scaleTime.ticks().map(tick => ({
      label: this.scaleTime.tickFormat()(tick),
      transform: this.tickTransform(tick)
    }));
  }

  get rangeLimit() {
    return this.scaleTime.range()[1];
  }

  private configureScaleTime(
    data: TimelineEvent[],
    timelineView: TimelineView,
    event: any
  ) {
    const scale = scaleTime()
      .domain([min(data, d => d.start), max(data, d => d.finish)])
      .range(this.getRange(timelineView));

    return event ? this.getTransformedScale(scale, event) : scale;
  }

  private getRange(view: TimelineView) {
    return this.orientation === Orientation.Vertical
      ? [view.top, view.bottom]
      : [view.left, view.right];
  }

  private tickTransform(tick: Date) {
    return this.orientation === Orientation.Vertical
      ? `translate(0,${this.mapToPixel(tick)})`
      : `translate(${this.mapToPixel(tick)}, 0)`;
  }
  private getTransformedScale(scale: ScaleTime<number, number>, event: any) {
    return this.orientation === Orientation.Vertical
      ? event.transform.rescaleY(scale)
      : event.transform.rescaleX(scale);
  }

  mapToPixel(domain: Date): number {
    return this.scaleTime(domain);
  }
}
