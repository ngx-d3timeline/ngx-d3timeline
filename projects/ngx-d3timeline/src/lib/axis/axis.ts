import { TickMark } from '../tick-mark/tick-mark';
import { Line } from './line';

export interface Axis {
  tickMarks: TickMark[];
  line: Line;
}
