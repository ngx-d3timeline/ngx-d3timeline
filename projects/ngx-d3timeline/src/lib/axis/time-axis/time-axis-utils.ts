import { TimelineView } from '../../view/timeline-view';
import { Orientation } from '../../orientation';
import { TimeScale } from '../../scale-types';
import { TimeAxisTickRenderer } from './time-axis-tick-renderer';
import { createSelector } from '../../selector/memoized-selector';
import { selectView, selectTimeOrientation } from '../../store/state';
import { selectTimeScale } from '../../store/timeline-selectors';
import { createAxis } from '../axis-utils';

export function createTimeAxis(
  view: TimelineView,
  orientation: Orientation,
  timeScale: TimeScale
) {
  return createAxis(view, orientation, timeScale, new TimeAxisTickRenderer());
}

export const selectTimeAxis = createSelector(
  [selectView, selectTimeOrientation, selectTimeScale],
  createTimeAxis
);
