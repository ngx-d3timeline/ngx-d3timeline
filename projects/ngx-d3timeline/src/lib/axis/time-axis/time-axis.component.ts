import { Component } from '@angular/core';
import { AxisService } from '../axis.service';

@Component({
  selector: '[ngx-d3timeline-time-axis]',
  template: `
    <svg:g
      class="time-axis-group"
      *ngIf="axisService.timeAxis$ | async as axis"
    >
      <svg:g *ngFor="let tick of axis.ticks" [attr.transform]="tick.transform">
        <svg:text dx="-7">
          {{ tick.label }}
        </svg:text>
        <svg:line class="tick-line" x2="-5"></svg:line>
      </svg:g>
      <svg:g ngx-d3timeline-axis-line [axisLine]="axis.axisLine"></svg:g>
    </svg:g>
  `,
  styles: [
    `
      .time-axis-group {
        text-anchor: end;
        dominant-baseline: central;
        font-size: 10px;
      }

      .tick-line {
        stroke: #000;
      }
    `
  ]
})
export class TimeAxisComponent {
  constructor(public axisService: AxisService) {}
}
