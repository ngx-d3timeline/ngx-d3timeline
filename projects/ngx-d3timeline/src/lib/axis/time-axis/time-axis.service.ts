import { Injectable } from '@angular/core';
import { ScalesService } from '../../scales.service';
import { ScaleTime } from 'd3-scale';
import { Orientation } from '../../orientation';
import { Axis } from '../axis';
import { Tick } from '../tick';
import { map } from 'rxjs/operators';
import { OptionsService } from '../../options.service';
import { AxisService } from '../axis.service';

@Injectable({ providedIn: 'root' })
export class TimeAxisService {
  vm$ = this.scalesService.scales$.pipe(
    map(scales =>
      this.createAxis(scales.scaleTime, scales.state.timelineOrientation)
    )
  );

  constructor(
    private scalesService: ScalesService,
    private optionsService: OptionsService,
    private axisService: AxisService
  ) {}

  private createAxis(
    scaleTime: ScaleTime<number, number>,
    timelineOrientation: Orientation
  ): Axis {
    return {
      ticks: this.getTicks(scaleTime, timelineOrientation),
      axisLine: this.axisService.getAxisLine(scaleTime, timelineOrientation)
    };
  }

  private getTicks(
    scale: ScaleTime<number, number>,
    orientation: Orientation
  ): Tick[] {
    return scale.ticks().map(value => ({
      label: scale.tickFormat()(value),
      transform: this.optionsService.getTranslation(scale(value), orientation)
    }));
  }
}
