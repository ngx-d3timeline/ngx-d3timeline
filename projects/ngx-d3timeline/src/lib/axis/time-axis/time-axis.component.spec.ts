import { TestBed, ComponentFixture } from '@angular/core/testing';
import { TimeAxisComponent } from './time-axis.component';
import { AxisService } from '../axis.service';
import { of } from 'rxjs';
import { Component, Input } from '@angular/core';
import { Line } from 'dist/ngx-d3timeline/lib/axis/line';
import { createLine } from '../line';
import { origin } from '../../point';

@Component({
  selector: '[ngx-d3timeline-line]',
  template: `
    <svg:g></svg:g>
  `
})
class FakeLineComponent {
  @Input() line: Line;
}

describe('TimeAxisComponent', () => {
  let fixture: ComponentFixture<TimeAxisComponent>;
  let axisService: AxisService;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [TimeAxisComponent, FakeLineComponent],
      providers: [{ provide: AxisService, useValue: { vm$: jest.fn() } }]
    });

    fixture = TestBed.createComponent(TimeAxisComponent);
    axisService = TestBed.inject(AxisService);
  });

  it('should not render if view model is null', () => {
    axisService.timeAxis$ = of(null);
    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const line = createLine(origin, { x: 10, y: 10 });

    axisService.timeAxis$ = of({
      ticks: [
        {
          label: 'tick 1',
          transform: 'translate(0, 10)',
          line
        },
        {
          label: 'tick 2',
          transform: 'translate(0, 20)',
          line
        }
      ],
      line: { x1: 0, x2: 10, y1: 1, y2: 0 }
    });

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
