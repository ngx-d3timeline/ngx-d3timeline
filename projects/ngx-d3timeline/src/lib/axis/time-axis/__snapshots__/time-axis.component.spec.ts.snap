// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeAxisComponent should not render if view model is null 1`] = `
<div
  id="root0"
  ngx-d3timeline-time-axis=""
>
  
</div>
`;

exports[`TimeAxisComponent should render correctly 1`] = `
<div
  id="root1"
  ngx-d3timeline-time-axis=""
>
  
  <g
    class="time-axis-group"
  >
    
    <g
      transform="translate(0, 10)"
    >
      <text
        dx="-7"
      >
         tick 1 
      </text>
      <line
        class="tick-line"
        x2="-5"
      />
    </g>
    <g
      transform="translate(0, 20)"
    >
      <text
        dx="-7"
      >
         tick 2 
      </text>
      <line
        class="tick-line"
        x2="-5"
      />
    </g>
    <g
      ngx-d3timeline-axis-line=""
    >
      <g />
    </g>
  </g>
</div>
`;
